staging/src/k8s.io/client-go/openapi3/root.go

```diff
@@ -0,0 +1,192 @@
+}
+
+// openapi3Root implements the Root interface, and encapsulates the
+// fields to retrieve, store the parsed OpenAPI V3 root document.
+type openapi3Root struct {
+	// Sync structure to allow only one download of root document.
+	once sync.Once
+	// REST client to retrieve the OpenAPI V3 documents.
+	client rest.Interface
+	// Parsed root OpenAPI V3 document mapping the GroupVersions
+	// to the OpenAPI V3 relative urls for subsequent retrieval of
+	// the GroupVersion schemas.
+	gvSchemas map[schema.GroupVersion]string
+	// Stores error in retrieval/parsing of root document.
+	err error
+}
+
+// Validate openapi3Root implements the Root interface.
+var _ Root = &openapi3Root{}
+
+// NewRoot returns a structure implementing the Root interface,
+// created with the passed rest client.
+func NewRoot(client rest.Interface) Root {
+	return &openapi3Root{client: client}
+}
+
+// GroupVersions returns all the GroupVersions for which the OpenAPI
+// V3 root document has a secondary OpenAPI V3 url. Returns an error
+// if there is a problem retrieving or parsing the OpenAPI V3 root
+// document.
+func (r *openapi3Root) GroupVersions() ([]schema.GroupVersion, error) {
+	r.retrieveV3Root()
+	if r.err != nil {
+		return nil, r.err
+	}
+	groupVersions := make([]schema.GroupVersion, len(r.gvSchemas))
+	for gv := range r.gvSchemas {
+		groupVersions = append(groupVersions, gv)
+	}
+	return groupVersions, nil
+}
+
+// GVSchemas returns the schemas for all the resources in a
+// GroupVersion as a pointer to a spec3.OpenAPI struct.
+// Returns an error for problems retrieving or parsing the root
+// document or GroupVersion OpenAPI V3 document.
+func (r *openapi3Root) GVSchemas(gv schema.GroupVersion) (*spec3.OpenAPI, error) {
+	openAPISchemaBytes, err := r.GVSchemasAsBytes(gv)
+	if err != nil {
+		return nil, err
+	}
+	// Unmarshal the downloaded Group/Version bytes into the spec3.OpenAPI struct.
+	var parsedV3Schema spec3.OpenAPI
+	err = json.Unmarshal(openAPISchemaBytes, &parsedV3Schema)
+	return &parsedV3Schema, err
+}
+
+// GVSchemasAsBytes returns the schemacs for all the resources in a
+// GroupVersion as unstructured bytes. Returns an error for
+// problems retrieving or parsing the root or GroupVersion
+// OpenAPI V3 documents.
+func (r *openapi3Root) GVSchemasAsBytes(gv schema.GroupVersion) ([]byte, error) {
+	r.retrieveV3Root()
+	if r.err != nil {
+		return nil, r.err
+	}
+	openAPIPath, found := r.gvSchemas[gv]
+	if !found {
+		return nil, fmt.Errorf("GroupVersion (%s), not found in open api V3 root", gv)
+	}
+	return r.client.Get().
+		RequestURI(openAPIPath).
+		SetHeader("Accept", runtime.ContentTypeJSON).
+		Do(context.TODO()).
+		Raw()
+}
+
+// retrieveV3Root retrieves and parses the root OpenAPI V3
+// document exactly once. Errors in this process are stored
+// in the "err" field.
+func (r *openapi3Root) retrieveV3Root() {
+	// Retrieve/parse the OpenAPI V3 root document only once.
+	r.once.Do(func() {
```

Comment:
This belongs to the downloaded, not the parser. and we can split into a cacheable downloader and a non-cacheable.